### PR TITLE
relayer/Address review. 

### PR DESCRIPTION
**Pull request by derpy-duck**
Only non-comment changes to the contracts being audited can be found in WormholeRelayerGovernance.sol:

We add two new errors: 
```
error ChainAlreadyRegistered(uint16 chainId, bytes32 registeredWormholeRelayerContract);
error InvalidDefaultDeliveryProvider(bytes32 defaultDeliveryProvider);
```

The first one is emitted if the current registered address is not the 0 address, and a registration is requested. This is so we can disallow re-registrations
The second is to disallow setting the default delivery provider to the 0 address (a simple sanity check)

and an event which we emit whenever the contract is upgraded
`event ContractUpgraded(address indexed oldContract, address indexed newContract);`

Additionally, we set the 'allowUnset' flag on parseAndCheckPayloadHeader in parseAndCheckRegisterDefaultDeliveryProviderVm to false. 

This is to disable the feature of having 'governance VAAs' where a default delivery provider is set, and no chainId is specified, meaning this VAA can register the default delivery provider onto any chain.

### CODE REVIEW

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol [hunk 1]
```diff
@@ -28,16 +28,19 @@ contract DeliveryProvider is DeliveryProviderGovernance, IDeliveryProvider {
...
-        Wei transactionFee =
-            quoteDeliveryOverhead(targetChain) + gasLimit.toWei(quoteGasPrice(targetChain));
-        Wei receiverValueCost = quoteAssetCost(targetChain, receiverValue);
-        nativePriceQuote = (
-            transactionFee.max(costOfProvidingFullGasLimit) + receiverValueCost
-        ).asLocalNative();
+        (uint16 buffer, uint16 denominator) = assetConversionBuffer(targetChain);
+        targetChainRefundPerUnitGasUnused = GasPrice.wrap(gasPrice(targetChain).unwrap() * (denominator) / (uint256(denominator) + buffer));
+        
+        TargetNative targetCostIfNoGasLimitUsed = gasLimit.toWei(targetChainRefundPerUnitGasUnused).asTargetNative();
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:34]: This calc does not appear to do what the var name suggests. As written it is "how much target chain currency would be refunded if no gas was used", not the cost to the user if no gas was used. 
The name implies the fixed overhead is included, but that is added later in the fn. 

Perhaps, max refund is more descriptive?

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol [hunk 2]
```diff
@@ -80,12 +83,10 @@ contract DeliveryProvider is DeliveryProviderGovernance, IDeliveryProvider {
         return sourceChainAmount.asNative().convertAsset(
             nativeCurrencyPrice(sourceChain),
             nativeCurrencyPrice(targetChain),
-            (buffer),
+            (bufferDenominator),
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:86]: Just confirming, the denom and num are meant to be swapped here correct?

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol [hunk 3]
```diff
@@ -28,16 +28,19 @@ contract DeliveryProvider is DeliveryProviderGovernance, IDeliveryProvider {
...
-        Wei receiverValueCost = quoteAssetCost(targetChain, receiverValue);
-        nativePriceQuote = (
-            transactionFee.max(costOfProvidingFullGasLimit) + receiverValueCost
-        ).asLocalNative();
+        (uint16 buffer, uint16 denominator) = assetConversionBuffer(targetChain);
+        targetChainRefundPerUnitGasUnused = GasPrice.wrap(gasPrice(targetChain).unwrap() * (denominator) / (uint256(denominator) + buffer));
+        
+        TargetNative targetCostIfNoGasLimitUsed = gasLimit.toWei(targetChainRefundPerUnitGasUnused).asTargetNative();
+        TargetNative targetCostIfFullGasLimitUsed = gasLimit.toWei(gasPrice(targetChain)).asTargetNative();
+        LocalNative costIfNoGasLimitUsed = quoteAssetCost(targetChain, targetCostIfNoGasLimitUsed);
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:36]: can this this use quoteGasCost?

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol [hunk 4]
```diff
@@ -113,28 +114,37 @@ contract DeliveryProvider is DeliveryProviderGovernance, IDeliveryProvider {
...
-        Wei gasPriceInSourceChainCurrency =
-            assetConversion(targetChain, gasPrice(targetChain).priceAsWei(), chainId());
-        require(gasPriceInSourceChainCurrency.unwrap() <= type(uint88).max, "Overflow");
-        return GasPrice.wrap(uint88(gasPriceInSourceChainCurrency.unwrap()));
+    function quoteGasPrice(uint16 targetChain) public view returns (GasPrice price) {
+        price = GasPrice.wrap(quoteGasCost(targetChain, Gas.wrap(1)).unwrap());
+        require(price.unwrap() <= type(uint88).max, "Overflow");
     }
 
+    error PriceIsZero(uint16 chain);
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:134]: Nit: move this to interface, or top of file

File: ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol [hunk 5]
```diff
@@ -93,7 +95,11 @@ abstract contract WormholeRelayerGovernance is WormholeRelayerBase, ERC1967Upgra
             foreignAddress;
     }
 
+    event ContractUpgraded(address indexed oldContract, address indexed newContract);
```
Comment (JoeHowarth) [inline ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol:98]: I believe ERC1967Upgrade already emits a similar event

**Review by JoeHowarth — CHANGES_REQUESTED** [review]

File: ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol (hunk 5 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol:98]: yup - this is the same event from TokenBridge, NFTBridge, CoreBridge, etc for consistency

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol (hunk 4 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:134]: done

File: ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol (hunk 1 above)
Comment (derpy-duck) [reply ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol:34]: fixed

**Review by JoeHowarth — APPROVED** [review]

**Pull request closed without merging**